proof/crefine/RISCV64/StateRelation_C.thy

```diff
@@ -351,18 +363,78 @@ definition
      \<and> ucast (tcbPriority atcb) = tcbPriority_C ctcb
      \<and> ucast (tcbMCP atcb) = tcbMCP_C ctcb
-     \<and> tcbTimeSlice atcb    = unat (tcbTimeSlice_C ctcb)
      \<and> cfault_rel (tcbFault atcb) (seL4_Fault_lift (tcbFault_C ctcb))
                   (lookup_fault_lift (tcbLookupFailure_C ctcb))
-     \<and> option_to_ptr (tcbBoundNotification atcb) = tcbBoundNotification_C ctcb"
+     \<and> option_to_ptr (tcbBoundNotification atcb) = tcbBoundNotification_C ctcb
+     \<and> option_to_ptr (tcbSchedContext atcb) = tcbSchedContext_C ctcb
+     \<and> option_to_ptr (tcbYieldTo atcb) = tcbYieldTo_C ctcb"
+
+definition crefill_size :: "sched_context_C \<Rightarrow> nat" where
+  "crefill_size sc \<equiv> if scRefillHead_C sc \<le> scRefillTail_C sc
+                        then unat (scRefillTail_C sc - scRefillHead_C sc + 1)
+                        else unat (scRefillTail_C sc + 1 + (scRefillMax_C sc - scRefillHead_C sc))"
+
+definition csched_context_relation :: "Structures_H.sched_context \<Rightarrow> sched_context_C \<Rightarrow> bool" where
+  "csched_context_relation asc csc \<equiv>
+       scPeriod asc = scPeriod_C csc
+     \<and> scConsumed asc = scConsumed_C csc
+     \<and> option_to_ptr (scTCB asc) = scTcb_C csc
+     \<and> option_to_ptr (scReply asc) = scReply_C csc
+     \<and> option_to_ptr (scNtfn asc) = scNotification_C csc
+     \<and> scBadge asc = scBadge_C csc
+     \<and> scSporadic asc = to_bool (scSporadic_C csc)
+     \<and> option_to_ptr (scYieldFrom asc) = scYieldFrom_C csc
+     \<and> scRefillMax asc = unat (scRefillMax_C csc)
+     \<and> scRefillHead asc = unat (scRefillHead_C csc)
+     \<and> scRefillCount asc = crefill_size csc"
+
+definition crefill_relation :: "refill \<Rightarrow> refill_C \<Rightarrow> bool" where
+  "crefill_relation ar cr \<equiv> rTime ar = rTime_C cr \<and> rAmount ar = rAmount_C cr"
+
+primrec dyn_array_list_rel ::
+  "('c :: c_type) typ_heap \<Rightarrow> ('a \<Rightarrow> 'c \<Rightarrow> bool) \<Rightarrow> 'a list \<Rightarrow> 'c ptr \<Rightarrow> bool"
+  where
+    "dyn_array_list_rel h rel [] p = True"
+  | "dyn_array_list_rel h rel (a # as) p
+      = (\<exists>v. h p = Some v \<and> rel a v \<and> dyn_array_list_rel h rel as (p +\<^sub>p 1))"
+
+definition sc_ptr_to_crefill_ptr :: "obj_ref \<Rightarrow> refill_C ptr" where
+  "sc_ptr_to_crefill_ptr p \<equiv> Ptr (p + of_nat sizeof_sched_context_t)"
+
+definition refill_buffer_relation ::
+  "(obj_ref \<rightharpoonup> kernel_object) \<Rightarrow> heap_raw_state \<Rightarrow> cghost_state \<Rightarrow> bool"
+  where
+  "refill_buffer_relation ah ch gs \<equiv>
+     let abs_sc_hp = map_to_scs ah;
+         crefill_hp = clift ch;
+         lens_hp = refill_buffer_lengths gs
+     in
+       dom abs_sc_hp = dom lens_hp
+     \<and> dom crefill_hp
+       = \<Union> ((\<lambda>p. set (map (\<lambda>i. sc_ptr_to_crefill_ptr p +\<^sub>p int i) [0..< the (lens_hp p)])) ` (dom abs_sc_hp))
```

Comment:
Looking at what Isabelle echoes back for the definition term, there is a slightly nicer way to write this (parsing to the exact same term):

```
∧ dom crefill_hp
       = (⋃p : dom abs_sc_hp. set  (map (λi. sc_ptr_to_crefill_ptr p +⇩p int i) [0..< the (lens_hp p)]))
```

Reply:
The indentation also seems a bit off here; the `∧` shouldn't be in the same column as the `in`. My preference would be for either the same column as the previous conjunct, or at the end of the previous line.